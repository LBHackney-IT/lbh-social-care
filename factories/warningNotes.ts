import { Factory } from 'fishery';

import { WarningNote } from 'types';

export const warningNoteFactory = Factory.define<WarningNote>(
  ({ sequence }) => ({
    id: sequence,
    type: 'Risk to Staff',
    startDate: new Date(2020, 11, 22),
    createdBy: 'Foo',
    disclosedWithIndividual: false,
    undisclosedDetails: 'nope',
    discussedWithManager: 'the big boss',
    discussedWithManagerDate: new Date(2020, 11, 12),
    reviewDate: new Date(2021, 1, 1),
    notes: 'a lot to talk about',
    status: 'open',
    reviews: [],
  })
);

export const mockedWarningNote = [
  warningNoteFactory.build({
    type: 'Risk to Adults',
    reviewedDate: new Date(2020, 11, 13),
    reviewedBy: 'Bar',
    disclosedDate: new Date(2020, 11, 12),
    disclosedHow: 'Written',
    disclosedWithIndividual: true,
    disclosedDetails: 'lorem ipsum',
    reviewDate: new Date(2020, 11, 31),
    notes: 'i am a note',
  }),
  warningNoteFactory.build(),
];
