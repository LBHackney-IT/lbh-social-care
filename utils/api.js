import { getQueryString } from 'utils/urls';

export const getInfiniteKey = (baseUrl, key, params) => (
  pageIndex,
  previousPageData
) => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
