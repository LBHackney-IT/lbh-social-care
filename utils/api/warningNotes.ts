import axios from 'axios';
import useSWR, { SWRResponse } from 'swr';

import type { WarningNote, ErrorAPI } from 'types';

export const useWarningNotes = (
  id: number
): SWRResponse<WarningNote[], ErrorAPI> =>
  useSWR(`/api/residents/${id}/warningnotes`);

export const useWarningNote = (
  warningNoteId: number
): SWRResponse<WarningNote, ErrorAPI> =>
  useSWR(`/api/warningnotes/${warningNoteId}`);

export const addWarningNote = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/warningnotes`, formData);
  return data;
};

export const updateWarningNote = async (
  noteId: number,
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/warningnotes/${noteId}`, formData);
  return data;
};
