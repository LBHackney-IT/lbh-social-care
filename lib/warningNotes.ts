import axios from 'axios';

import { mockedWarningNote } from 'factories/warningNotes';
import { WarningNote } from 'types';

const { ENDPOINT_API, AWS_KEY } = process.env;

const headers = { 'x-api-key': AWS_KEY };

export const getWarningNotesByResident = async (
  personId: number
): Promise<WarningNote[] | []> => {
  const { data } = await axios.get(
    `${ENDPOINT_API}/residents/${personId}/warningnotes`,
    {
      headers,
    }
  );
  return data.warningNotes;
};

export const getWarningNote = async (
  warningNoteId: number
): Promise<WarningNote | undefined> => {
  return await new Promise((resolve) =>
    setTimeout(
      () => resolve(mockedWarningNote.find(({ id }) => id === warningNoteId)),
      1000
    )
  );
};

export const addWarningNote = async (
  params: Record<string, unknown>
): Promise<void> => {
  await axios.post(`${ENDPOINT_API}/warningnotes`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};

export const patchWarningNote = async (
  personId: number,
  params: Record<string, unknown>
): Promise<void> => {
  await axios.patch(`${ENDPOINT_API}/warningnotes/${personId}`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};
