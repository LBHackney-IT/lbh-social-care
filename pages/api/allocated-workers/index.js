import * as HttpStatus from 'http-status-codes';

import { getAllocatedWorkers } from 'utils/server/allocatedWorkers';
import { isAuthorised } from 'utils/auth';

export default async (req, res) => {
  if (!isAuthorised({ req })) {
    return res.status(HttpStatus.UNAUTHORIZED).send('Auth cookie missing.');
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getAllocatedWorkers(req.query);
        if (data?.ascAllocations?.length > 0) {
          res.status(HttpStatus.OK).json(data);
        } else {
          res.status(HttpStatus.NOT_FOUND).json('Allocated Workers Not Found');
        }
      } catch (error) {
        console.log('Allocated Workers get error:', error.response.data);
        error?.response?.status === HttpStatus.NOT_FOUND
          ? res.status(HttpStatus.NOT_FOUND).json('Allocated Workers Not Found')
          : res
              .status(HttpStatus.INTERNAL_SERVER_ERROR)
              .json('Unable to get the Allocated Workers');
      }
      break;

    // case 'POST':
    //   try {
    //     const data = await addAllocatedWorker(req.query.id, req.body);
    //     res.status(HttpStatus.OK).json(data);
    //   } catch (error) {
    //     console.log('Allocated Workers post error:', error.response.data);
    //     res
    //       .status(HttpStatus.INTERNAL_SERVER_ERROR)
    //       .json('Unable to post Allocated Workers');
    //   }
    //   break;

    default:
      res.status(HttpStatus.BAD_REQUEST).json('Invalid request method');
      console.log(res.status);
  }
};
