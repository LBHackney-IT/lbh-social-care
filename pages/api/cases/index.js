import { postResidentCase } from 'utils/server/residents';
import { isAuthorised } from 'utils/auth';

export default async (req, res) => {
  if (!isAuthorised({ req })) {
    return res.status(401).send('Auth cookie missing.');
  }
  switch (req.method) {
    case 'POST':
      try {
        const { data } = await postResidentCase(req.query.id, req.body);
        res.setHeader('Content-Type', 'application/json');
        res.end(JSON.stringify(data));
      } catch (error) {
        console.log(error.status);
        console.log('Cast post error:', error);
        res.status(500).json('Unable to post cast');
      }
      break;

    default:
      res.status(400).json('Invalid request method');
      console.log(res.status);
  }
};
