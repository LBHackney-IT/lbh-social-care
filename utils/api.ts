import { getQueryString } from 'utils/urls';

interface PageData {
  nextCursor?: string;
  [key: string]: unknown;
}

export const getInfiniteKey = (
  baseUrl: string,
  key: string,
  params: Record<string, unknown>
) => (pageIndex: number, previousPageData: PageData): string | null => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
