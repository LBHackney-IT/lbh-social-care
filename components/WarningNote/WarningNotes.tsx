import cx from 'classnames';
import Link from 'next/link';

import { useWarningNotes } from 'utils/api/warningNotes';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import { WarningNote } from 'types';

import styles from './WarningNotes.module.scss';

export interface Props {
  personId: number;
  notes: WarningNote[];
}

export const WarningBox = ({ notes, personId }: Props): React.ReactElement => {
  return (
    <div
      className={cx('govuk-error-summary', styles.container)}
      aria-labelledby="warning-note-title"
      role="alert"
      tabIndex={-1}
    >
      <h2 className="govuk-error-summary__title" id="warning-note-title">
        WARNING NOTE
      </h2>
      <div>
        <div className="govuk-error-summary__body">
          {notes.map((note) => (
            <div key={note.id} className={styles.note}>
              <dl className={styles.noteDetails}>
                {note.noteType && (
                  <>
                    <dt>Type</dt>
                    <dd>{note.noteType}</dd>
                  </>
                )}
                {note.startDate && (
                  <>
                    <dt>Start date</dt>
                    <dd>
                      {note.reviewedDate
                        ? new Date(note.reviewedDate).toLocaleDateString(
                            'en-GB'
                          )
                        : new Date(note.startDate).toLocaleDateString(
                            'en-GB'
                          )}{' '}
                      <i>created by</i> {note.createdBy}
                    </dd>
                  </>
                )}
                {note.reviewDate && (
                  <>
                    <dt>Review date</dt>
                    <dd>
                      {new Date(note.reviewDate).toLocaleDateString('en-GB')}{' '}
                    </dd>
                  </>
                )}
                {note.nextReviewDate && (
                  <>
                    <dt>Next review date</dt>
                    <dd>
                      {new Date(note.nextReviewDate).toLocaleDateString(
                        'en-GB'
                      )}{' '}
                    </dd>
                  </>
                )}
              </dl>
              <Link href={`/people/${personId}/warning-notes/${note.id}`}>
                <a className="govuk-link govuk-link-underline">Review / end</a>
              </Link>
            </div>
          ))}
          <p>For further information see Warning Note in Records History</p>
        </div>
      </div>
    </div>
  );
};

const WarningNotes = ({ id }: { id: number }): React.ReactElement | null => {
  const { data: warningNotes, error } = useWarningNotes(id);
  if (error) {
    return <ErrorMessage />;
  }
  if (!warningNotes || warningNotes.length === 0) {
    return null;
  }
  return <WarningBox notes={warningNotes} personId={id} />;
};

export default WarningNotes;
