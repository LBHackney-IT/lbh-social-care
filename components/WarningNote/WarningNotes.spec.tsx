import { render } from '@testing-library/react';

import * as warningNotes from 'utils/api/warningNotes';
import WarningNotes from './WarningNotes';

jest.mock('next/router', () => ({
  useRouter: () => ({
    asPath: 'path',
    push: jest.fn(),
  }),
}));

jest.mock('utils/api/warningNotes', () => ({
  useWarningNotes: jest.fn(),
}));

describe(`useWarningNotes`, () => {
  jest.spyOn(warningNotes, 'useWarningNotes').mockImplementation(() => ({
    data: [
      {
        id: 123,
        type: 'Risk to Adults',
        createdDate: new Date(2020, 12, 12),
        createdBy: 'Foo',
        reviewedDate: new Date(2020, 12, 13),
        reviewedBy: 'Bar',
      },
      {
        id: 321,
        type: 'Risk to Adults',
        createdDate: new Date(2020, 12, 12),
        createdBy: 'Foo',
        reviewedDate: new Date(2020, 12, 13),
        reviewedBy: 'Bar',
      },
      {
        id: 234,
        type: 'Risk to Staff',
        createdDate: new Date(2020, 12, 22),
        createdBy: 'Foo',
      },
    ],
    mutate: jest.fn(),
    revalidate: jest.fn(),
    isValidating: false,
  }));

  const props = {
    id: '123',
  };

  it('should render properly', () => {
    const { asFragment } = render(<WarningNotes {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
