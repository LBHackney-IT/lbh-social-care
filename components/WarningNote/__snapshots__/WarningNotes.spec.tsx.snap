// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWarningNotes should render properly 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="warning-note-title"
    class="govuk-error-summary container"
    role="alert"
    tabindex="-1"
  >
    <h2
      class="govuk-error-summary__title"
      id="warning-note-title"
    >
      WARNING NOTE
    </h2>
    <div>
      <div
        class="govuk-error-summary__body"
      >
        <div
          class="note"
        >
          <dl
            class="noteDetails"
          >
            <dt>
              Type
            </dt>
            <dd>
              Risk to Adults
            </dd>
            <dt>
              Start Date
            </dt>
            <dd>
              13/12/2020 
              <i>
                created by
              </i>
               Foo
            </dd>
            <dt>
              Review Date
            </dt>
            <dd>
              31/12/2020 
            </dd>
          </dl>
          <a
            class="govuk-link govuk-link-underline"
            href="/people/123/warning-notes/1"
          >
            Review / end
          </a>
        </div>
        <div
          class="note"
        >
          <dl
            class="noteDetails"
          >
            <dt>
              Type
            </dt>
            <dd>
              Risk to Staff
            </dd>
            <dt>
              Start Date
            </dt>
            <dd>
              22/12/2020 
              <i>
                created by
              </i>
               Foo
            </dd>
            <dt>
              Review Date
            </dt>
            <dd>
              01/02/2021 
            </dd>
          </dl>
          <a
            class="govuk-link govuk-link-underline"
            href="/people/123/warning-notes/2"
          >
            Review / end
          </a>
        </div>
        <p>
          For further information see Warning Note in Records History
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
